File: Java/1-50/14.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.math.BigInteger;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        String input = reader.readLine();
        String[] inputs = input.split(" ");
        long n = Long.parseLong(inputs[0]);
        long k = Long.parseLong(inputs[1]);

        long output;
        if (n == 0) {
            output = 1;
        } else {
            output = modularPow(k + 1, n, 1000000007);
        }

        System.out.println(output);
    }

    public static long modularPow(long baseNumber, long exponent, long modulus) {
        BigInteger result = BigInteger.ONE;
        BigInteger base = BigInteger.valueOf(baseNumber);
        BigInteger mod = BigInteger.valueOf(modulus);

        while (exponent > 0) {
            if (exponent % 2 == 1) {
                result = result.multiply(base).mod(mod);
            }
            base = base.multiply(base).mod(mod);
            exponent /= 2;
        }

        return result.longValue();
    }
}

File: Kotlin/1-50/14.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.math.BigInteger

fun main() {
    val reader = BufferedReader(InputStreamReader(System.`in`))
    val input = reader.readLine()
    val inputs = input.split(" ")
    val n = inputs[0].toLong()
    val k = inputs[1].toLong()

    val output: Long
    if (n == 0L) {
        output = 1L
    } else {
        output = modularPow(k + 1, n, 1000000007)
    }

    println(output)
}

fun modularPow(baseNumber: Long, exponent: Long, modulus: Long): Long {
    var result = BigInteger.ONE
    var base = BigInteger.valueOf(baseNumber)
    val mod = BigInteger.valueOf(modulus)

    var exp = exponent
    while (exp > 0) {
        if (exp % 2 == 1L) {
            result = result.multiply(base).mod(mod)
        }
        base = base.multiply(base).mod(mod)
        exp /= 2
    }

    return result.toLong()
}
